### DISCUSSION TITLE
Request / Reply Wrapper for Presentation API with Node.js Server

### DISCUSSION THREAD

**Discussion by youngbrett** [iTwin/itwinjs-core#2369] [Q&A] [answered]
Hi,

In the Unity example app, I'd like to get info on Revit properties using the Presentation API.  I added the dependencies for presentation-backend and initialized it with a ruleset in Main.ts.  I follow the other request / reply wrappers and can replicate them, but I don't see where the presentation API is called.  I'm following the reply to [this discussion](https://gist.github.com/grigasp/32131ec63c5132e7599c387e9a835330).  Any help on how to do this with node.js?

Thank you.

**Comment by grigasp** [comment]
I hope @mgooding can comment regarding Unity specifics. 

As for the Presentation APIs, in your given example the API to get the properties is called on [line 127](https://gist.github.com/grigasp/32131ec63c5132e7599c387e9a835330#file-main-ts-L127-L135). 

I would also recommend to have a look at this example: https://www.itwinjs.org/sandboxes/grigas/Element%20Properties%20Loader. It's a frontend sample, but the Presentation APIs on the frontend and the backend are very similar. The sample demonstrates usage of a different Presentation API called `getElementProperties` - it's easier to use compared to `getContent` as it requires less parameters and the return type is less verbose and easier to parse.

**Reply by youngbrett** [reply]
Thanks, @grigasp.  I'm trying to replace Unity Reflect, which has access to Revit parameters as a metadata class; if I have these parameters, I can ditch Reflect for a much better (flexible, faster, more reliable) stack.  But thing on a thing on a thing on a thing on a thing. (Revit on iTwin on Protobuf on Node.js on Unity.) 

Looking more at it, I think the call needs to be made in [ProtobufRPCServer.ts](https://github.com/imodeljs/imodel-unity-example/blob/master/node/src/ProtobufRpcServer.ts), putting the results into a flexible data type that can be parsed on receipt of the reply wrapper.

@mgooding - are there examples of other uses of the Node.js / Protobuf scheme with iTwins, which I could look at?

**Reply by mgooding** [reply]
> Looking more at it, I think the call needs to be made in ProtobufRPCServer.ts, putting the results into a flexible data type that can be parsed on receipt of the reply wrapper.

Yes, this is where I'd put it. Prompted by your question, I'm working on adding this to the official example. I'll put up a PR later today.

**Reply by youngbrett** [reply]
@mgooding - very much appreciate it.  Will show you how I'm using it soon.  For reference, [here's what the Reflect metadata looks like in Unity](https://forum.unity.com/proxy.php?image=https%3A%2F%2Fi.imgur.com%2FXwBUXZH.png&hash=eeedfa973807e0f3f3904aae82cc7a28).

**Reply by mgooding** [reply]
The Presentation example is now integrated into imodel-unity-example. You can see the necessary changes here: https://github.com/imodeljs/imodel-unity-example/commit/c8e8d46dda7e36243c4ebacec32ce533db35c56a.

Please let me know if you have any questions or suggestions. Thanks!

**Comment by youngbrett** [comment, marked as answer]
Thanks for all the help!  Pulled and it looks great.
